Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp

```diff
@@ -558,17 +572,56 @@ void RemoteLayerTreeEventDispatcher::renderingUpdateComplete()
 }
 
 #if ENABLE(THREADED_ANIMATION_RESOLUTION)
+void RemoteLayerTreeEventDispatcher::lockForAnimationChanges()
+{
+    ASSERT(isMainRunLoop());
+    m_animationsLock.lock();
+}
+
+void RemoteLayerTreeEventDispatcher::unlockForAnimationChanges()
+{
+    ASSERT(isMainRunLoop());
+    m_animationsLock.unlock();
+    startOrStopDisplayLink();
+}
+
 void RemoteLayerTreeEventDispatcher::animationsWereAddedToNode(RemoteLayerTreeNode& node)
 {
+    ASSERT(isMainRunLoop());
+    assertIsHeld(m_animationsLock);
     auto effectStack = node.takeEffectStack();
     ASSERT(effectStack);
     m_effectStacks.set(node.layerID(), effectStack.releaseNonNull());
 }
 
 void RemoteLayerTreeEventDispatcher::animationsWereRemovedFromNode(RemoteLayerTreeNode& node)
 {
+    ASSERT(isMainRunLoop());
+    assertIsHeld(m_animationsLock);
     m_effectStacks.remove(node.layerID());
 }
+
+void RemoteLayerTreeEventDispatcher::updateAnimations()
+{
+    ASSERT(!isMainRunLoop());
+    Locker lock { m_animationsLock };
+
+    // FIXME: Rather than using 'now' at the point this is called, we
+    // should probably be using the timestamp of the (next?) display
+    // link update or vblank refresh.
+    auto secondsSinceEpoch = MonotonicTime::now().secondsSinceEpoch();
+
+    auto effectStacks = std::exchange(m_effectStacks, { });
+    for (auto& [layerID, effectStack] : effectStacks) {
+        effectStack->applyEffectsFromScrollingThread(secondsSinceEpoch);
+
+        // We can clear the effect stack if it's empty, but the previous
+        // call to applyEffects() is important so that the base values
+        // were re-applied.
+        if (effectStack->hasEffects())
+            m_effectStacks.set(layerID, WTFMove(effectStack));
```

Comment:
I was confused by the "clear then repopulate". It would be clearer to just remove ones that are no longer active?